darts/models/forecasting/nbeats.py

```diff
@@ -30,48 +31,60 @@ class _TrendGenerator(nn.Module):
-        self.T = nn.Parameter(
-            torch.stack([(torch.arange(output_dim) / output_dim)**i for i in range(expansion_coefficient_dim)], 1),
-            False)
+
+        # basis is of size (expansion_coefficient_dim, target_length)
+        basis = torch.stack([(torch.arange(target_length) / target_length)**i for i in range(expansion_coefficient_dim)], 
+                            dim=1).T
+
+        self.basis = nn.Parameter(basis, requires_grad=False)
+
 
     def forward(self, x):
-        return torch.matmul(x, self.T.T)
+        return torch.matmul(x, self.basis)
 
 
 class _SeasonalityGenerator(nn.Module):
 
     def __init__(self,
-                 output_dim):
+                 target_length):
         super(_SeasonalityGenerator, self).__init__()
-        half_minus_one = int(output_dim / 2 - 1)
-        cos_vectors = [torch.cos(torch.arange(output_dim) * 2 * np.pi * i) for i in range(1, half_minus_one + 1)]
-        sin_vectors = [torch.sin(torch.arange(output_dim) * 2 * np.pi * i) for i in range(1, half_minus_one + 1)]
-        self.S = nn.Parameter(torch.stack([torch.ones(output_dim)] + cos_vectors + sin_vectors, 1),
-                              False)
+        half_minus_one = int(target_length / 2 - 1)
+        cos_vectors = [torch.cos(torch.arange(target_length) * 2 * np.pi * i) for i in range(1, half_minus_one + 1)]
+        sin_vectors = [torch.sin(torch.arange(target_length) * 2 * np.pi * i) for i in range(1, half_minus_one + 1)]
+        basis = torch.stack([torch.ones(target_length)] + cos_vectors + sin_vectors, dim=1).T
+
+        self.basis = nn.Parameter(basis, requires_grad=False)
 
     def forward(self, x):
-        return torch.matmul(x, self.S.T)
+        return torch.matmul(x, self.basis)
 
 
 class _Block(nn.Module):
 
     def __init__(self,
                  num_layers: int,
                  layer_width: int,
+                 nr_params: int,
                  expansion_coefficient_dim: int,
                  input_chunk_length: int,
                  target_length: int,
                  g_type: GTypes):
         """ PyTorch module implementing the basic building block of the N-BEATS architecture.
 
+        The blocks produce outputs of size (target_length, nr_params); i.e.
+        "one vector per parameter". The parameters are predicted only for forecast outputs.
+        Backcast outputs are in the original "domain".
+
         Parameters
         ----------
         num_layers
             The number of fully connected layers preceding the final forking layers.
         layer_width
             The number of neurons that make up each fully connected layer.
+        nr_params
+            The number of parameters of the likelihood (or 1 if no likelihood is used)
         expansion_coefficient_dim
             The dimensionality of the waveform generator parameters, also known as expansion coefficients.
             Only used if `generic_architecture` is set to `True`.
```

Comment:
I just realized this is not true, because the trend module also needs the `expansion_coefficient_dim`, which it uses as degree for the polynomial basis.

```suggestion
            Used in the generic architecture and the trend module of the interpretable architecture, where it determines the degree of the polynomial basis.
```

Reply:
I guess you might want to split this into 2 lines or something..

Reply:
Actually I think this parameter is used only when `generic_architecture` is `True`. When it is `False, we are using `trend_polynomial_degree + 1` (line 354) for the trend (and `2 * int(input_chunk_length / 2 - 1) + 1` for the seasonality).

Reply:
Yeah that's right but at line 354 we are passing `trend_polynomial_degree + 1` as `expansion_coefficient_dim` parameter to the stack module. The stack module does not accept a separate `trend_polynomial_degree` parameter. So I the description is correct for the `_NBEATSModule` and `NBEATSModel` classes, but from `_Stack` downward it is not correct anymore because the trend module also uses `expansion_coefficient_dim` to determine the basis dimension. Does that make sense?

Reply:
Right, makes sense 👍